### PR TITLE
PUD-1202: Add audit support for remaining recall database fields

### PR DESCRIPTION
**Pull request by philsegal**

### CODE REVIEW

File: src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/RecallAuditComponentTest.kt [hunk 1]
```diff
@@ -53,7 +56,108 @@ class RecallAuditComponentTest : ComponentTestBase() {
   }
 
   @Test
-  fun `get audit summary for a recall that has been booked`() {
+  fun `get lastReleaseDate (LocalDate value) audit history`() {
+    val savedRecall = authenticatedClient.bookRecall(BookRecallRequest(nomsNumber, FirstName("Brian"), null, LastName("Badgering")))
+    val recallId = savedRecall.recallId
+
+    val updatedRecall = authenticatedClient.updateRecall(recallId, UpdateRecallRequest(lastReleaseDate = LocalDate.now()))
```
Comment (AndyNewton) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/RecallAuditComponentTest.kt:63]: Would be interesting to add some parameterized test examples - though also I haven't tried for you.

**Review by AndyNewton — APPROVED** [review]
Approved with just a couple of QQs.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/RecallAuditService.kt [hunk 2]
```diff
@@ -58,28 +62,47 @@ class RecallAuditService(
...
 
+  private fun FieldName.getNameFromPath(): String = this.value.split(".").last()
+
   private fun translateValue(updatedValue: String, fieldName: FieldName): Any =
-    when (fieldName) {
-      FieldName("contraband") -> booleanValue(updatedValue)
+    when (getRecallFieldAndPath(fieldName).first.type) {
+      java.lang.Boolean::class.java -> booleanValue(updatedValue)
+      LocalDate::class.java -> LocalDate.parse(updatedValue)
+      OffsetDateTime::class.java -> OffsetDateTime.parse(updatedValue.replace(" ", "T"))
```
[the remark this answers is not in the document]
Comment (philsegal) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/RecallAuditService.kt:103]: The issue here is that the precision seems to be different between local and ci builds. Not looked into why, or whether we can fix that precision via the database TIMESTAMPZ(x). 
Now I have something working, I will look into that.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/RecallAuditComponentTest.kt (hunk 1 above)
Comment (philsegal) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/RecallAuditComponentTest.kt]: Done

File: src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/RecallAuditComponentTest.kt [hunk 3]
```diff
@@ -53,7 +56,108 @@ class RecallAuditComponentTest : ComponentTestBase() {
...
+
+    val updatedRecall = authenticatedClient.updateRecall(
+      recallId,
+      UpdateRecallRequest(
+        sentenceLength = Api.SentenceLength(3, 0, 0),
+        sentenceDate = LocalDate.now(),
+        licenceExpiryDate = LocalDate.now(),
+        indexOffence = "Offence 1",
+        sentencingCourt = CourtId("ACCRYC"),
+        sentenceExpiryDate = LocalDate.now()
```
[the remark this answers is not in the document]
Comment (philsegal) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/RecallAuditComponentTest.kt]: Done

File: src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/RecallAuditComponentTest.kt [hunk 4]
```diff
@@ -19,54 +23,43 @@ import uk.gov.justice.digital.hmpps.managerecallsapi.domain.PrisonId
...
     val recallId = savedRecall.recallId
 
-    val updatedRecall = authenticatedClient.updateRecall(recallId, UpdateRecallRequest(lastReleaseDate = LocalDate.now()))
+    val updatedRecall = authenticatedClient.updateRecall(recallId, updateRequest)
 
-    val auditList = authenticatedClient.auditForField(recallId, "lastReleaseDate")
+    val auditList = authenticatedClient.auditForField(recallId, fieldName.value)
 
     assertThat(auditList.size, equalTo(1))
-    assertThat(auditList[0].updatedValue, equalTo(LocalDate.now().toString()))
```
Comment (AndyNewton) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/RecallAuditComponentTest.kt]: FYI I quite like to see one non-parameterized test - which can be more complete/specific/explicit - alongside a parameterized test for breadth of coverage.

**Review by AndyNewton — APPROVED** [review]

**Pull request merged**